### PR TITLE
pid1: various fixes around TTY resets

### PR DESCRIPTION
**Pull request by poettering**
Fixes: #29714

### CODE REVIEW

File: src/core/execute.c [hunk 1]
```diff
@@ -150,14 +169,10 @@ void exec_context_tty_reset(const ExecContext *context, const ExecParameters *p)
...
-                (void) reset_terminal_fd(fd, true);
-
-        if (p && p->stdin_fd >= 0) {
-                unsigned rows = context->tty_rows, cols = context->tty_cols;
+                (void) reset_terminal_fd(fd, /* switch_to_text= */ true);
 
-                (void) exec_context_tty_size(context, &rows, &cols);
-                (void) terminal_set_size_fd(p->stdin_fd, path, rows, cols);
-        }
+        if (fd >= 0)
```
Comment (YHNdnzj) [inline src/core/execute.c]: fd is guaranteed to be >= 0 here IIUC?

**Review by YHNdnzj — APPROVED** [review]

File: src/core/execute.c [hunk 2]
```diff
@@ -99,27 +99,46 @@ const char *exec_context_tty_path(const ExecContext *context) {
         return "/dev/console";
 }
 
-int exec_context_tty_size(const ExecContext *context, unsigned *ret_rows, unsigned *ret_cols) {
+void exec_context_determine_tty_size(
```
Comment (yuwata) [inline src/core/execute.c]: Not a matter, but this can be static.

**Review by yuwata — APPROVED** [review]

**Comment by yuwata** [comment]
Downgrading the green label, but feel free to merge this as is.

File: src/core/execute.c (hunk 2 above)
Comment (poettering) [reply src/core/execute.c]: fixed.

File: src/core/exec-invoke.c [hunk 3]
```diff
@@ -413,26 +413,24 @@ static int setup_input(
...
+                tty_fd = acquire_terminal(tty_path,
+                                          i == EXEC_INPUT_TTY_FAIL  ? ACQUIRE_TERMINAL_TRY :
+                                          i == EXEC_INPUT_TTY_FORCE ? ACQUIRE_TERMINAL_FORCE :
+                                                                      ACQUIRE_TERMINAL_WAIT,
+                                          USEC_INFINITY);
+                if (tty_fd < 0)
+                        return tty_fd;
 
-                r = terminal_set_size_fd(fd, exec_context_tty_path(context), rows, cols);
+                r = exec_context_apply_tty_size(context, tty_fd, tty_path);
```
Comment (YHNdnzj) [inline src/core/exec-invoke.c:429]: `tty_fd` is leaked if this call fails

File: src/core/exec-invoke.c [hunk 4]
```diff
@@ -413,26 +413,24 @@ static int setup_input(
...
+                int tty_fd;
 
-                fd = acquire_terminal(exec_context_tty_path(context),
-                                      i == EXEC_INPUT_TTY_FAIL  ? ACQUIRE_TERMINAL_TRY :
-                                      i == EXEC_INPUT_TTY_FORCE ? ACQUIRE_TERMINAL_FORCE :
-                                                                  ACQUIRE_TERMINAL_WAIT,
-                                      USEC_INFINITY);
-                if (fd < 0)
-                        return fd;
+                tty_path = exec_context_tty_path(context);
```
Comment (YHNdnzj) [inline src/core/exec-invoke.c]: Maybe `assert_se`?

**Comment by poettering** [comment]
made the suggested bugfixes/tweaks, ptal

**Review by yuwata — APPROVED** [review]

**Pull request merged**